src/ui/App.swift

```diff
@@ -233,33 +237,169 @@ class App: AppCenterApplication {
     }
 
     func showUiOrCycleSelection(_ shortcutIndex: Int, _ forceDoNothingOnRelease_: Bool) {
+        showUiOrCycleSelectionWithSource(shortcutIndex, forceDoNothingOnRelease_, .unknown)
+    }
+    
+    func showUiOrCycleSelectionWithSource(_ shortcutIndex: Int, _ forceDoNothingOnRelease_: Bool, _ eventSource: EventSource) {
+        // Atomic operation protection - prevent concurrent calls
```

Comment:
I don't think it's possible to have concurrent calls of this method. It should always be called from `DispatchQueue.main`. Calls should be sequentials.

There are cases of concurrency in the apps. There are always main queue + another queue. Once we get executed on the main queue, concurrency issues should be gone since that code should always be called from `DispatchQueue.main`, not from other queues. If it is called from other queues, it's a bug and should be fixed.